ReCAPTCHA on User Lost Password Page

Godoy0722 opened the pull request:
*Description*: Add ReCAPTCHA on the `userLostPassword.tpl` page. Solves the issue #6984.

jonasraoni commented on classes/template/PKPTemplateManager.inc.php:
```diff
@@ -205,12 +205,15 @@ function initialize($request) {
 					'https://www.recaptcha.net/recaptcha/api.js?hl=' . substr(AppLocale::getLocale(),0,2),
 					[
-						'contexts' => ['frontend-user-register', 'frontend-user-registerUser'],
+						'contexts' => ['frontend-user-register', 'frontend-user-registerUser', 'frontend-login-lostPassword'],
```
This should be a conditional include, see this for reference: https://github.com/pkp/pkp-lib/pull/6541/files#diff-8100c829b3c8e2b0f5bbf2d00cbc29f5b67dfa0b7183132bfa8694c7d6e476a7R202-R218

Godoy0722 replied:
Gotcha. I'll send a new commit with these updates. Thanks @jonasraoni !

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
+		$captchaEnabled = Config::getVar('captcha', 'captcha_on_password_reset') && Config::getVar('captcha', 'recaptcha');
+		if ($captchaEnabled) {
+			import('lib.pkp.classes.form.Form');
+			$form = new Form('');
```
```suggestion
			$form = new Form();
```

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
 		$userDao = DAORegistry::getDAO('UserDAO'); /** @var UserDAO $userDao */
 		$user = $userDao->getUserByEmail($email);
 
+		$rateLimitingEnabled = true;
```
I think this should be added to a configuration, and perhaps disabled by default to avoid breaking changes.
Then a user can opt-in for both the recaptcha and the rate limiter.

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
+			$session = $sessionManager->getUserSession();
+
+			// Store reset requests with timestamps in the session
+			$resetRequests = $session->getSessionVar('passwordResetRequests') ?: [];
```
Hmm, if we store the retries on the session, then I think it's not very useful, as clearing the session cookie will be enough to skip the check.

I think this piece of code should be removed or moved to another issue, there's a relevant discussion here: https://github.com/pkp/pkp-lib/discussions/11084

Godoy0722 replied:
Gotcha. I'll remove this piece of code for now.

jonasraoni commented on templates/frontend/pages/userLostPassword.tpl:
```diff
@@ -39,6 +39,14 @@
 				</label>
 			</div>
+
+			{if $captchaEnabled}
+			<div class="captcha">
+				<div class="pkp_form_locale_field">
+					{$reCaptchaHtml}
+				</div>
+			</div>
+			{/if}
```
Just to follow the pattern of other places:

```suggestion
			{if $reCaptchaHtml}
			<div class="captcha">
				<div class="pkp_form_locale_field">
					{$reCaptchaHtml}
				</div>
			</div>
			{/if}
```

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
+
+			if (!$captchaValidator->isValid()) {
+				$templateMgr->assign([
+					'captchaEnabled' => true,
```
```suggestion
```

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
+			}
+
+			// Check if this email has a recent request (within the last 10 minutes)
+			if (isset($resetRequests[$email]) && ($currentTime - $resetRequests[$email] < 600)) {
```
If the user was blocked, I think it's good to inform that a new request can be made in "X minutes"

jonasraoni commented on pages/login/LoginHandler.inc.php:
```diff
@@ -169,12 +180,71 @@ function requestResetPassword($args, $request) {
+
+			// Remove entries older than 1 hour (3600 seconds)
+			foreach ($resetRequests as $resetEmail => $timestamp) {
+				if ($currentTime - $timestamp > 3600) {
```
If the user is allowed to try again after 10 minutes, then I think we can clear the entries also after 10 minutes.

Godoy0722 commented:
Hello @jonasraoni, just a heads-up that I implemented your change requests, as far as I also removed the rate limiter as you previously mentioned. Please let me know if I have something else to do in this issue and I'll update it. :smile:

asmecher commented:
Thanks, @Godoy0722, that looks good! This will also need to be added to the themes we maintain that override the changed templates, or else they'll stop working if the option is enabled.
